### PR TITLE
Add Mii selector applet (appletEd)

### PR DESCRIPTION
**Pull request by phijor**
Lately, @s0r00t and I have been reverse engineering the Mii Selector applet. This is a working implementation of how to configure and launch it. I documented what we found out [in the wiki](https://www.3dbrew.org/wiki/Mii_Selector).

Please criticize any of the code and/or documentation. I tried to keep the implementation close to how the interface to `swkbd` and `errDisp` is implemented, but right now this code is far less verbose than that of those applets.

### CODE REVIEW

**Comment by s0r00t** [comment]
Making a usage example could be a swift addition. 
Also, why didn't you use booleans instead of chars in the struct for flags?

File: libctru/include/3ds/applets/miiselector.h [hunk 1]
```diff
@@ -0,0 +1,111 @@
+/**
+ * @file miiselector.h
+ * @brief Mii Selector Applet (appletEd).
+ */
+
+#pragma once
+
+/// @cond INTERNAL
+#define _assert_struct_size(s, size) \
+	_Static_assert(sizeof(s) == (size), #s " is not of size " #size "!")
```
Comment (mtheall) [inline libctru/include/3ds/applets/miiselector.h]: This limits you to C11. I would do it like this:
```
/// @cond INTERNAL
#ifdef __cplusplus
#if __cplusplus >= 201103L
#define _assert_struct_size(s, size) \
	static_assert(sizeof(s) == (size), #s " is not of size " #size "!")
#endif
#else
#if __STDC_VERSION__ >= 201112L
#define _assert_struct_size(s, size) \
	_Static_assert(sizeof(s) == (size), #s " is not of size " #size "!")
#endif
#endif
/// @endcond

File: libctru/include/3ds/applets/miiselector.h [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+/**
+ * @brief Mii selector context
+ *
+ * Configure how to launch the applet in @ref MiiSelectorContext.config. After
+ * returning, the applet has written data to @ref MiiSelectorContext.ret.
+ */
+typedef union {
+	MiiSelectorConf config; ///< Configuration passed to applet
+	MiiSelectorReturn ret;  ///< Data written by applet after return
+} MiiSelectorContext;
```
Comment (mtheall) [inline libctru/include/3ds/applets/miiselector.h]: Instead of using a union, I would have miiSelectorLaunch() take two parameters, one for the input MiiSelectorConf and one for the output MiiSelectorReturn.

File: libctru/include/3ds/applets/miiselector.h [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+
+/**
+ * @brief Launch the Mii selector library applet
+ */
+Result miiSelectorLaunch(MiiSelectorContext *ctx);
```
Comment (mtheall) [inline libctru/include/3ds/applets/miiselector.h]: From above, this becomes:

    Result miiSelectorLaunch(const MiiSelectorConf *conf, MiiSelectorReturn *returnbuf);

File: libctru/include/3ds/applets/miiselector.h [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+
+/**
+ * @brief Verifies that the Mii data returned from the applet matches its
+ * checksum
+ *
+ * @param returnbuffer Buffer filled by Mii selector applet
+ *
+ * @return `true` if `returnbuf->checksum` is the same as the one computed from `returnbuf`
+ */
+bool miiSelectorChecksumIsValid(MiiSelectorReturn *returnbuf);
```
Comment (mtheall) [inline libctru/include/3ds/applets/miiselector.h]: Use `const MiiSelectorReturn *returnbuf`

File: libctru/source/applets/miiselector.c [hunk 5]
```diff
@@ -0,0 +1,42 @@
+#include <3ds/types.h>
+#include <3ds/services/apt.h>
+
+#include <3ds/applets/miiselector.h>
+
+Result miiSelectorLaunch(MiiSelectorContext *ctx)
```
Comment (mtheall) [inline libctru/source/applets/miiselector.c]: From above, this would look like the following:

```
Result miiSelectorLaunch(const MiiSelectorConf *conf, MiiSelectorReturn *returnbuf);
{
	union {
		MiiSelectorConf config;
		MiiSelectorReturn ret;
	} ctx;

	memcpy(&ctx.config, conf, sizeof(MiiSelectorConf));
	ctx.config.magic = MIISELECTOR_MAGIC;

	Result ret = aptLaunchLibraryApplet(APPID_APPLETED, &ctx->config, sizeof(MiiSelectorConf), 0);
	if(R_SUCCEEDED(ret) && returnbuf)
		memcpy(returnbuf, &ctx.ret, sizeof(MiiSelectorReturn));

	return ret;
}
```

**Comment by phijor** [comment]
Adding an example to [devkitPro/3ds-examples ](https://github.com/devkitPro/3ds-examples) would be a good idea, indeed.

As it stands right now, I do not know if `sizeof(bool) == sizeof(u8)`. The structure critically depends on having a fixed memory layout, that's why I did not dare making the flags `bool`s.

File: libctru/include/3ds/applets/miiselector.h (hunk 1 above)
Comment (fincs) [inline libctru/include/3ds/applets/miiselector.h]: Why is this here anyway? It doesn't belong in this header file (maybe types.h?)

File: libctru/source/applets/miiselector.c (hunk 5 above)
Comment (phijor) [inline libctru/source/applets/miiselector.c]: I wasn't really happy with the `union` too; this is definitely a better approach. Thanks for the input, will implement!

File: libctru/include/3ds/applets/miiselector.h (hunk 1 above)
Comment (phijor) [inline libctru/include/3ds/applets/miiselector.h]: @mtheall: Sorry, forgot that this was C11 only. What should this macro expand to if neither of your suggested conditions are met, `((void) 0)`?

@fincs: Of all the exisiting files I would guess `types.h` to be the most fitting, I don't really like here either (just didn't want to touch other files rn). Are there any other places in ctrulib that could benefit from using this macro or similar ones?

Comment (fincs) [reply libctru/include/3ds/applets/miiselector.h]: Can we get rid of this assert_struct_size thing? I don't think it's necessary when it's already verified to be correct.

Comment (phijor) [reply libctru/include/3ds/applets/miiselector.h]: Done.
